Userland/Utilities/sed.cpp

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2022, Eli Youngs <eli.m.youngs@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/GenericLexer.h>
+#include <AK/Vector.h>
+#include <LibCore/ArgsParser.h>
+#include <LibCore/Stream.h>
+#include <LibCore/System.h>
+#include <LibMain/Main.h>
+#include <LibRegex/RegexMatcher.h>
+#include <LibRegex/RegexOptions.h>
+
+struct SubstitutionCommand {
+    Regex<PosixExtended> regex;
+    StringView replacement;
+    PosixOptions options;
+};
+
+static ErrorOr<SubstitutionCommand> parse_command(StringView command)
+{
+    char const generic_error_message[] = "Incomplete substitution command";
+
+    auto lexer = GenericLexer(command);
+
+    auto address = lexer.consume_until('s');
+    if (!address.is_empty())
+        warnln("sed: Addresses are currently ignored");
+
+    if (!lexer.consume_specific('s'))
+        return Error::from_string_literal(generic_error_message);
+
+    if (lexer.is_eof())
+        return Error::from_string_literal(generic_error_message);
+
+    auto delimiter = lexer.consume();
+    if (delimiter == '\n' || delimiter == '\\')
+        return Error::from_string_literal("\\n and \\ cannot be used as delimiters.");
+
+    auto pattern = lexer.consume_until(delimiter);
+    if (pattern.is_empty())
+        return Error::from_string_literal("Substitution patterns cannot be empty.");
+
+    if (!lexer.consume_specific(delimiter))
+        return Error::from_string_literal(generic_error_message);
+
+    auto replacement = lexer.consume_until(delimiter);
+
+    // According to Posix, "s/x/y" is an invalid substitution command.
+    // It must have a closing delimiter: "s/x/y/"
+    if (!lexer.consume_specific(delimiter))
+        return Error::from_string_literal("The substitution command was not properly terminated.");
+
+    PosixOptions const options = PosixOptions(PosixFlags::Global | PosixFlags::SingleMatch);
+
+    auto flags = lexer.consume_all();
+    if (!flags.is_empty())
+        warnln("sed: Flags are currently ignored");
+
+    return SubstitutionCommand { Regex<PosixExtended> { pattern }, replacement, options };
+}
+
+ErrorOr<int> serenity_main(Main::Arguments args)
+{
+    TRY(Core::System::pledge("stdio rpath"));
+
+    Core::ArgsParser args_parser;
+
+    StringView command_input;
+    Vector<StringView> filepaths;
+
+    args_parser.add_positional_argument(command_input, "Command", "command_input", Core::ArgsParser::Required::Yes);
+    args_parser.add_positional_argument(filepaths, "File", "file", Core::ArgsParser::Required::No);
+
+    args_parser.parse(args);
+
+    auto command = TRY(parse_command(command_input));
+
+    if (filepaths.is_empty())
+        filepaths = { "-"sv };
```

Comment:
Could you move this above the command parsing (it doesn't depend on it) and unveil yourself to the given files at this point? If stdin is the only necessary file, you can unveil away the entire file system which is very nice.

The point here is that full sed expression syntax is pretty complicated, and its parser may easily be exploitable in a way that an attacker gets read access to other files. Unveil is a simple mitigation mechanism preventing read access to all files except the ones on the command line, which means that sed with untrusted patterns is safer to do.

Reply:
Unfortunately, I think this is a more involved process. A substitution command like `s/x/y/wabc` can require us to create and write to `abc`, regardless of whether the input comes from `stdin`, so we'd need to unveil those paths as well. I tried implementing this, but it looks like unveil throws an error on relative paths, so I'll either need to update `unveil` itself, or add additional logic to canonicalize any paths used in substitution commands. Right now there's only one substitution command per run, but once we support more of sed's features, there will be more places for us to handle reading and writing to arbitrary files, so I don't think adding the unveil logic here will work as a long-term solution.

Would you consider this a requirement for merging this PR?